pkg/auth/auth_oidc_test.go

```diff
@@ -0,0 +1,631 @@
+	server.HandleFunc("/.well-known/openid-configuration", m.handleDiscovery)
+	return nil
+}
+
+func (m *mockOIDCProvider) endpointConfig() oauth2.Endpoint {
+	return oauth2.Endpoint{
+		AuthURL:  m.issuer + "/auth",
+		TokenURL: m.issuer + "/token",
+	}
+}
+
+func startMockProvider(t *testing.T) (*mockOIDCProvider, *url.URL, func() error) {
+	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+
+	providerURL := &url.URL{Scheme: "http", Host: listener.Addr().String()}
+	provider := newMockOIDCProvider(providerURL, privKey)
+	handler := http.NewServeMux()
+	provider.attachEndpoints(handler)
+
+	go http.Serve(listener, handler)
+
+	return provider, providerURL, listener.Close
+}
+
+func Test_oidcAuth_login(t *testing.T) {
+	encryptionKey := []byte(randomString(32))
+	authnKey := []byte(randomString(64))
+
+	oidcProvider, providerURL, closePort := startMockProvider(t)
+	defer closePort()
+
+	tests := []struct {
+		name       string
+		token      *oauth2.Token
+		wantUserID string
+		wantErr    bool
+	}{
+		{
+			name:    "no token",
+			wantErr: true,
+		},
+		{
+			name:       "id_token only",
+			token:      addIDToken(&oauth2.Token{}, oidcProvider.signPayload(`{"sub":"testuser","exp":`+strconv.FormatInt(time.Now().Add(5*time.Minute).Unix(), 10)+`}`)),
+			wantUserID: "testuser",
+		},
+		{
+			name:       "ID and refresh tokens",
+			token:      addIDToken(&oauth2.Token{RefreshToken: "testing-refresh"}, oidcProvider.signPayload(`{"sub":"testuser","exp":`+strconv.FormatInt(time.Now().Add(5*time.Minute).Unix(), 10)+`}`)),
+			wantUserID: "testuser",
+		},
+		{
+			name:    "invalid ID token signature",
+			token:   addIDToken(&oauth2.Token{RefreshToken: "testing-refresh"}, oidcProvider.signPayload(`{"sub":"testuser","exp":`+strconv.FormatInt(time.Now().Add(5*time.Minute).Unix(), 10)+`}`)+"abbaba"),
+			wantErr: true,
+		},
+		{
+			name:    "id_token missing sub claim",
+			token:   addIDToken(&oauth2.Token{RefreshToken: "testing-refresh"}, oidcProvider.signPayload(`{"exp":`+strconv.FormatInt(time.Now().Add(5*time.Minute).Unix(), 10)+`}`)),
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			o, err := newOIDCAuth(
+				context.Background(),
+				sessions.NewSessionStore(authnKey, encryptionKey, true, "/"),
+				&oidcConfig{
+					getClient: func() *http.Client {
+						return http.DefaultClient
+					},
+					issuerURL:             providerURL.String(),
+					clientID:              testClientID,
+					secureCookies:         true,
+					constructOAuth2Config: testOAuth2ConfigConstructor,
+				},
+			)
+			require.NoError(t, err)
+
+			// The OIDC provider apparently needs some time to load the keys from the
+			// JWKs URI, and so the provider object that we get is not 100% ready
+			// to verify keys when we get it. Wow.
+			time.Sleep(20 * time.Millisecond)
```

Comment:
The truth appears to be a bit more complicated. Token verification actually wouldn't use the context that's pushed to `provider.Verify()` to update the keys, but it uses the context it was constructed with.

However, the async cache implementation we've got will have long cancelled that context for the first item retrieved during initialization thanks to the `PollUntilContextTimeout` logic used there.